Improved Number.prototype.toString for radix other than 10

chqrlie opened the pull request:
- fix the conversions for integers and exact fractions
- approximate approach for other cases.
- bypass floating point conversions for JS_TAG_INT values
- avoid divisions for base 10 integer conversions
- fixes #242

chqrlie committed:
Improved Number.prototype.toString for radix other than 10

- fix the conversions for integers and exact fractions
- approximate approach for other cases.
- bypass floating point conversions for JS_TAG_INT values
- avoid divisions for base 10 integer conversions
- fixes #242

chqrlie commented:
I purposely did not remove the radix argument from `js_dtoa` and `js_dtoa1` because we should change the implementation to avoid relying on the C library `snprintf` which has some issues on Windows.

bnoordhuis commented:
You will want to update v8.txt in the top-level directory but otherwise nice work.

chqrlie committed:
Improved Number.prototype.toString for radix other than 10

- fix the conversions for integers and exact fractions
- approximate approach for other cases.
- bypass floating point conversions for JS_TAG_INT values
- avoid divisions for base 10 integer conversions
- fixes #242

chqrlie commented:
I updated `v8.txt`.  Why does `v8.sh` tee the output to the console? Just showing the diff would suffice and avoid verbose listings.

chqrlie committed:
Improved Number.prototype.toString for radix other than 10

- fix the conversions for integers and exact fractions
- approximate approach for other cases.
- bypass floating point conversions for JS_TAG_INT values
- avoid divisions for base 10 integer conversions
- fixes #242

bnoordhuis commented on quickjs.c:
```diff
@@ -11238,6 +11245,80 @@ static JSValue js_dtoa(JSContext *ctx,
+        }
+        *ptr2 = '\0';
+        if (frac * radix >= radix / 2) {
+            char nine = '0' + radix - 1;
```
For my understanding, this is sort of incorrect when radix > 10 but works okay in practice? I tried to deduce if you could get an accidental match in e.g. base-18 or base-36 (because then `nine` is 'A' or 'S') but I'm not sure.

I suppose it could also lead to somewhat unexpected results on EBCDIC systems but :shrug:

chqrlie replied:
This is incorrect indeed for bases larger than 10. I shall post a fix.  The code should be
```
    char nine = digits[radix - 1];
```
I am using an array for the digit characters, so EBCDIC should work fine :)

bnoordhuis reviewed (approved):
> Why does v8.sh tee the output to the console?

It made it easier for me to debug things.

chqrlie committed:
Improved Number.prototype.toString for radix other than 10

- fix the conversions for integers and exact fractions
- approximate approach for other cases.
- bypass floating point conversions for JS_TAG_INT values
- avoid divisions for base 10 integer conversions
- fixes #242

bnoordhuis reviewed (approved):
LGTM. Nice work.

chqrlie commented:
> LGTM. Nice work.

Thank you.  For completeness, this implementation is partial, it is not fully precise for the last fractional digit or two because of precision issues. I will post a multi-precision version when I get it done.
